longhorndev/dev.go

```diff
@@ -239,45 +258,16 @@ func (d *LonghornDevice) PrepareUpgrade() error {
 }
 
 func (d *LonghornDevice) FinishUpgrade() (err error) {
-	if d.frontend == "" {
-		return nil
-	}
-
-	stopCh := make(chan struct{})
-	socketError := d.WaitForSocket(stopCh)
-	select {
-	case err = <-socketError:
-		if err != nil {
-			logrus.Errorf("error waiting for the socket %v", err)
-			err = errors.Wrapf(err, "error waiting for the socket")
-		}
-		break
-	}
-	close(stopCh)
-	close(socketError)
-
-	if err != nil {
+	if err := d.start(false); err != nil {
 		return err
 	}
 
-	if err = d.ReloadSocketConnection(); err != nil {
-		return err
-	}
-
-	bsOpts := fmt.Sprintf("size=%v", d.size)
-	scsiDev, err := iscsidev.NewDevice(d.name, d.GetSocketPath(), "longhorn", bsOpts)
-	if err != nil {
-		return err
-	}
-	d.scsiDevice = scsiDev
-
-	return nil
+	// TODO: Need to fix `ReloadSocketConnection` since it doesn't work for frontend `FrontendTGTISCSI`.
+	return d.ReloadSocketConnection()
```

Comment:
Should it be reload first then start device? If reload failed, then we cannot start the device.

Reply:
But we need to wait for the socket ready first before reloading... That's why I choose to start the device first.

Reply:
Oh, probably need the d.getDev() for the reload socket call?

Reply:
Don't need to merge the two parts of logic into start() though. You can leave wait for socket here still.

Reply:
Now we can reload first then init the device later.